Add ability to configure custom parameter restrictions

jdmarsh opened the pull request:
We require the ability to configure a more restricted set of parameters from the python front end, specifically around the polynomial size. I've created this PR which is a very naive approach just to achieve the desired functionality, it would be great if someone could let me know how best to fit this functionality into the current API design and I'll get it tidied up.

BourgerieQuentin commented:
That's something we wanted to do, so that great to see that PR and an actual need for exposing a search space restriction. I added @aPere3 as reviewer and we will do deeper suggestion/review soon.

aPere3 commented on compilers/concrete-compiler/compiler/include/concretelang/Support/V0Parameters.h:
```diff
@@ -80,6 +80,10 @@ const concrete_optimizer::Encoding DEFAULT_ENCODING =
 const bool DEFAULT_CACHE_ON_DISK = true;
 const uint32_t DEFAULT_CIPHERTEXT_MODULUS_LOG = 64;
 const uint32_t DEFAULT_FFT_PRECISION = 53;
+const uint64_t DEFAULT_LOG2_POLYNOMIAL_SIZE_MIN = std::numeric_limits<uint64_t>::min();
+const uint64_t DEFAULT_LOG2_POLYNOMIAL_SIZE_MAX = std::numeric_limits<uint64_t>::max();
+const uint64_t DEFAULT_GLWE_DIMENSION_MIN = std::numeric_limits<uint64_t>::min();
+const uint64_t DEFAULT_GLWE_DIMENSION_MAX = std::numeric_limits<uint64_t>::max();
```
This encodes a kind of soft `Optional` when it arrives in `default_cpu` in `config.rs`. I think an explicit optional would be better here.

aPere3 commented on compilers/concrete-compiler/compiler/include/concretelang/Support/V0Parameters.h:
```diff
@@ -110,6 +114,13 @@ struct CompositionRule {
 const std::vector<CompositionRule> DEFAULT_COMPOSITION_RULES = {};
 const bool DEFAULT_COMPOSABLE = false;
 
+struct ParameterRestrictions {
+  uint64_t log2_polynomial_size_min;
+  uint64_t log2_polynomial_size_max;
+  uint64_t glwe_dimension_min;
+  uint64_t glwe_dimension_max;
+};
+
```
Do we need a separate struct here ? Can't `concrete_optimizer::ParameterRestrictions` be used directly ?

aPere3 replied:
Also, I would use optionals here.

aPere3 commented on compilers/concrete-compiler/compiler/lib/Bindings/Python/CompilerAPIModule.cpp:
```diff
@@ -923,6 +923,14 @@ void mlir::concretelang::python::populateCompilerAPISubmodule(
            [](CompilationOptions &options, int security_level) {
              options.optimizerConfig.security = security_level;
            })
+      .def("set_parameter_restrictions",
```
Just so that you know, `std::optional` are properly recognized by pybind. if you use them in the signature, it will work as expected.

aPere3 commented on compilers/concrete-optimizer/concrete-optimizer/src/optimization/config.rs:
```diff
@@ -29,12 +37,21 @@ pub struct SearchSpace {
-            .log2_polynomial_size
-            .as_vec();
-        let glwe_dimensions: Vec<u64> = DEFAUT_DOMAINS.glwe_pbs_constrained.glwe_dimension.as_vec();
+    pub fn default_cpu(parameter_restrictions: ParameterRestrictions) -> Self {
+        let mut glwe_log_polynomial_sizes_range = DEFAUT_DOMAINS.glwe_pbs_constrained.log2_polynomial_size;
+        glwe_log_polynomial_sizes_range.start = std::cmp::max(glwe_log_polynomial_sizes_range.start,
+            parameter_restrictions.log2_polynomial_size_min);
+        glwe_log_polynomial_sizes_range.end = std::cmp::min(glwe_log_polynomial_sizes_range.end,
+            parameter_restrictions.log2_polynomial_size_max);
+        let glwe_log_polynomial_sizes: Vec<u64> = glwe_log_polynomial_sizes_range.as_vec();
+
+        let mut glwe_dimensions_range = DEFAUT_DOMAINS.glwe_pbs_constrained.glwe_dimension;
+        glwe_dimensions_range.start = std::cmp::max(glwe_dimensions_range.start,
+            parameter_restrictions.glwe_dimension_min);
+            glwe_dimensions_range.end = std::cmp::min(glwe_dimensions_range.end,
+            parameter_restrictions.glwe_dimension_max);
+        let glwe_dimensions: Vec<u64> = glwe_dimensions_range.as_vec();
+
```
From what I understand, the use of `std::cmp::min` and `std::cmp::max` is a way to softly encode an `std::optional`. I think an explicit one would be better here.

BourgerieQuentin replied:
That's also what I understand, but the thing that bother me is that it restrict the search space to the intersection of the search space of our cpu backend and the parameter_restriction, it would be better to let all the freedom to define the search space as mainly the restriction of the parameter space is due of implementation details of tfhe-rs (if well remember only the polynomial size supported in our fft) and hardware provider may have a larger search space.
To resume I think I would prefer a way to define a custom search space instead of only a restrictions of the cpu one.

aPere3 reviewed (changes requested):
Hello @jdmarsh !

Thanks for this PR 😃 !

It looks mostly OK for me. The only thing I would like to see changed, is the fact that you use the min and max representant of u64 and `std::cmp::min/max` to softly encode an optional. I think something explicit would be better. I made a few comments here and there about that.

Also, if you have the time, it may be worth expanding the feature to the other domains, as this is a trivial extension to your PR, and it would make the feature complete.

Cheers,

Alex

BourgerieQuentin commented on compilers/concrete-optimizer/concrete-optimizer/src/optimization/config.rs:
```diff
@@ -76,9 +93,9 @@ impl SearchSpace {
         match processing_unit {
-            config::ProcessingUnit::Cpu => Self::default_cpu(),
+            config::ProcessingUnit::Cpu => Self::default_cpu(parameter_restrictions),
             config::ProcessingUnit::Gpu {
```
If we want to implement that as a parameter restriction should also be ported on gpu parameter selection, I think the easy way would be to implement a `fn apply_restriction(Self, ParameterRestrictions) -> Self` and apply that to the whole expression.

BourgerieQuentin reviewed (commented):
I added my comments, we will sync with @aPere3 tomorrow and see in which direction we want to go.

jdmarsh commented:
@aPere3 

To address your comments:
- I've extended the support to all of the `ParameterDomains` members.
- Added a separate default for CPU/GPU `glwe_pbs_constrained` ranges as a different range was hard coded into the SearchSpace gpu functions.
- In the pybind, I've used std::optional as the input parameters.
- I've created python functions in `compilation_options.py` to call into these with some parameter checking.
- I was unable to use std::optional to pass the values from the C++ side to the rust side. As we use cxx bridge, support for Optional is not implemented ([https://github.com/dtolnay/cxx?tab=readme-ov-file#builtin-types](https://github.com/dtolnay/cxx?tab=readme-ov-file#builtin-types)) I decided to use SharedPtr as it's nullable and behaves the same way.
- Since we use SharedPtr I've changed the `Option` args to references so that the memory ownership stays in the C++ side.

Please let me know if there are any errors/tidying/changes you'd like me to make.

Thanks,
James

jdmarsh commented:
@aPere3 @BourgerieQuentin Have you had a chance to take a look at the latest changes?

aPere3 commented:
Hey @jdmarsh 😃 ,

Thanks for getting back to us and sorry for the delay. The PR looks good enough for merge. As @BourgerieQuentin mentioned, some work I do at the moment have some overlap, but it's gonna take a bit of time before it gets merge so I prefer not to block. 

I'll rebase your commits, sign them (which is needed now), and manage the conflicts for you, so that you don't need to do any more changes to the PR. Hopefully it will be merged by the end of the day.

If you need it, you will have access to the feature in the next nightly. 

Cheers ! 

Alex

jdmarsh commented:
Thanks @aPere3 I really appreciate your help!
